### DISCUSSION TITLE
Predictable parquet format schema is important with hive_partitioning (varchar -> bigint not desirable)

### DISCUSSION THREAD

**Discussion by geraldodev** [duckdb/duckdb#15476] [General]
Hi,

I'm generating Parquet files by connecting DuckDB to a PostgreSQL database. I understand that it uses a mechanism of binary copying. I've noticed that sometimes varchar fields are converted to bigint in the target Parquet file when using the COPY command.

Since I'm using Hive partitioning, having a consistent schema across partitions is crucial to avoid issues where one partition has a different schema from another.

I've tried creating an in-memory DuckDB table with the same PostgreSQL schema and loading the data there, but even so, when reading from the in-memory DuckDB table, varchar fields are still being read as bigint.

It would be helpful if we could specify the field types explicitly with the COPY command. Hive partitioning is very useful, but maintaining a uniform schema with it is essential.

What do you think ?

Regards,

Geraldo Lopes de Souza

**Comment by soerenwolfers** [comment]
You should create an issue with a minimal working example. In this MWE, you should make clear whether the problem is in the PG->DuckDB step, in the DuckDB->Parquet step, or in the HivePartitionKeys->DuckDB step and omit all other, unnecessary, steps.

**Comment by geraldodev** [comment]
Hi, 

I've made further tests, 
I'm reading data through duckdb from postgres with copy pg.table to 'parquet (format parquet, "
                                                       partition-expression " ,"
                                                       "OVERWRITE true,"
                                                       "FILENAME_PATTERN \"" nome-lower "_{i}\""
                                                       ",COMPRESSION zstd"
                                                       ",WRITE_PARTITION_COLUMNS true"
                                                       ");"
                                              )

I've inserted an artificial record on postgres table with value 'V' for varchar  fields to force the values to be recognized as varchar.
In addition to that I've changed the copy command to use a select
(select * from pg.table
 order by partition-field-1 desc) 
 to force the varchar record to come first (in case of partial analysis of the content)
 
 the problem is that each partition gets its own schema reflecting the data data is being written to it, so the 'V' record goes to its own partition with the type of the field as VARCHAR, while the same field in other partitions are recognized as BIGINT
 
 A simple way to see this in action is: consider a field in postgres table, month varchar(2). it will be a partition selector.
month: '01', '02','03','10','11','12'

The first 9 partition will get varchar type, from '10' on it'll get bigint type.

That's not useful when dealing with hivemind partitioning because that forces one to select month::varchar as month in all queries that have this field, and depending on the values you get, with the new parquet files you are going to have query breakages.

So the safe way would be to  ::varchar alias all fields supposed to be varchar.

Parquet with hivemind partition are a match in heaven to accumulate data that are produced periodically (in partitions). A predictable schema is very important. So I propose in the case of the copy to command from an attached table,  the schema of the source table to be considered instead of the content of the data.

Regards,

Geraldo Lopes de Souza

**Reply by soerenwolfers** [reply]
You're going to have to boil it down and break it up into individual issues that you think need fixing, see my previous comment.
